### ISSUE TITLE
Trait as a singleton

### ISSUE THREAD

**Issue by xerial** [wvlet/airframe#21]
`@PreDestory` and `@PostConstruct` are useful for marking methods for initializing or shuting down objects, however, when these annotations are used in traits, the owner of these methods becomes ambiguous. Here is an example of adding `@PreDestroy` to a trait `Service` can be mixed in other trait A and B:
```
trait Service {
  val x = bind[X]
  val y = bind[Y]

  @PreDestory
  def close {
     x.close()
     y.close()
  }
}

trait A extends Service
trait B extends Service

val session = 
  newDesign
  .bind[X].toSingleton
  .bind[Y].toSingleton
  .newSession

val a = session.build[A]
val b = session.build[B]
```

In this case, `@PreDestory` methods are duplicated in trait A and B, so the owner of `@PreDestroy` is A and B; So close needs to be called twice, but the actual intention is making X and Y as singletons, and the close method for X and Y should be called only once.

To fix this situation `Service` needs to be a singleton so that we can associate `@PreDestroy` hook to `Service` trait and the design have a singleton binding for `Service`. 

A problem is trait is usually for injecting the same code to multiple locations, so duplicating `@PreDestroy` code to A and B is a natural behavior. We may need to have #22 (Singleton annotation) to clarify the semantics.

**Comment by xerial** [comment]
A workaround is adding a wrapper XYService to make sure the scope of `@PreDestory`:

```
trait XYService {
   val x = bind[X]
   val y = bind[Y]
  
  @PreDestory
  def close {
     x.close()
     y.close()
  }
}

trait XYServiceModule {
   val service = bind[XYService]
}

trait A extends XYServiceModule {
  import service._
}
trait B extends XYServiceModule {
  import service._
}

val session = 
  newDesign
  .bind[XYService].toSingleton
  .newSession
```

**Comment by xerial** [comment]
Introduced `bindSingleton[X]` in #23

**Comment by xerial** [comment]
Closed by #23
